pertpy/tools/_distance_tests.py

```diff
@@ -0,0 +1,224 @@
+
+    def test_xy(self, adata: AnnData, groupby: str, contrast: str, verbose: bool = True) -> pd.DataFrame:
+        """Run permutation test for metrics that take x and y."""
+        distance = Distance(self.metric, self.obsm_key)
+        groups = adata.obs[groupby].unique()
+        if contrast not in groups:
+            raise ValueError(f"Contrast group {contrast} not found in {groupby} of adata.obs.")
+        fct = track if verbose else lambda x: x
+        emb = adata.obsm[self.obsm_key]
+        res = []
+
+        # Generate the null distribution
+        for _i in fct(range(self.n_perms)):
+            # per perturbation, shuffle with control and compute e-distance
+            df = pd.DataFrame(index=groups, columns=["distance"], dtype=float)
+            for group in groups:
+                if group == contrast:
+                    continue
+                # Shuffle the labels of the groups
+                mask = adata.obs[groupby].isin([group, contrast])
+                labels = adata.obs[groupby].values[mask]
+                shuffled_labels = np.random.permutation(labels)
+                idx = shuffled_labels == group
+
+                X = emb[mask][idx]  # shuffled group
+                Y = emb[mask][~idx]  # shuffled contrast
+                dist = distance(X, Y)
+
+                df.loc[group, "distance"] = dist
+            res.append(df.sort_index())
+
+        # Generate the empirical distribution
+        for group in groups:
+            if group == contrast:
+                continue
+            X = emb[adata.obs[groupby] == group]
+            Y = emb[adata.obs[groupby] == contrast]
+            df.loc[group, "distance"] = distance(X, Y)
+
+        # Evaluate the test
+        # count times shuffling resulted in larger distance
+        results = np.array(pd.concat([r["distance"] - df["distance"] for r in res], axis=1) > 0, dtype=int)
+        n_failures = pd.Series(np.clip(np.sum(results, axis=1), 1, np.inf), index=df.index)
+        pvalues = n_failures / self.n_perms
+
+        # Apply multiple testing correction
+        significant_adj, pvalue_adj, _, _ = multipletests(pvalues.values, alpha=self.alpha, method=self.correction)
+
+        # Aggregate results
+        tab = pd.DataFrame(
+            {
+                "distance": df["distance"],
+                "pvalue": pvalues,
+                "significant": pvalues < self.alpha,
+                "pvalue_adj": pvalue_adj,
+                "significant_adj": significant_adj,
+            },
+            index=df.index,
+        )
+
+        # Set the contrast group
+        tab.loc[contrast, "distance"] = 0
+        tab.loc[contrast, "pvalue"] = 1
+        tab.loc[contrast, "significant"] = False
+        tab.loc[contrast, "pvalue_adj"] = 1
+        tab.loc[contrast, "significant_adj"] = False
+
+        return tab
+
+    def test_precomputed(self, adata: AnnData, groupby: str, contrast: str, verbose: bool = True) -> pd.DataFrame:
```

Comment:
The naming is also not thaaat great maybe. But not terrible.

In any case, please add a full docstring.